File: rendering_api/metal/src/core/presentation_queue/metal_presentation_queue.cpp
#define GN_TEXTURE_NO_TEXTURE_ID
#include <core/presentation_queue/gryphn_device_presentation_details.h>
#include <core/presentation_queue/gryphn_presentation_queue.h>
#include "core/devices/metal_output_devices.h"
#include "core/metal_instance.h"
#include "core/textures/metal_texture.h"
#include "QuartzCore/CAMetalLayer.h"
#include "core/sync_objects/metal_semaphore.h"
#include "bridge/metal_bridge.h"
#include "metal_presentation_queue.h"

gnOutputDevice* mltGetOutputDevice(gnPlatformPresentationQueue* presentaionQueue) {
    return presentaionQueue->outputDevice;
}

#define MAXIMUM_DRAWABLES 3

GN_EXPORT gnDevicePresentationDetails gnGetDevicePresentationDetailsFn(const gnPhysicalOutputDevice& physicalOutputDevice) {
    return { MAXIMUM_DRAWABLES, MAXIMUM_DRAWABLES };
}

GN_EXPORT gnReturnCode gnPresentationQueueGetNextImageAsyncFn(gnPresentationQueue& presentationQueue, const gnSyncSemaphore& semaphore, gnUInt* imageIndex) {
    mtlObjectCSetContentViewsLayer(
        presentationQueue.presentationQueue->outputDevice->outputDevice->instance->instance->window,
        presentationQueue.presentationQueue->layer);

    presentationQueue.presentationQueue->currentDrawableIndex++;
    presentationQueue.presentationQueue->currentDrawableIndex %= MAXIMUM_DRAWABLES;
    *imageIndex = presentationQueue.presentationQueue->currentDrawableIndex;

    presentationQueue.presentationQueue->currentDrawable = presentationQueue.presentationQueue->layer->nextDrawable();

    dispatch_semaphore_signal(semaphore.semaphore->semaphore);
    return GN_SUCCESS;
}

GN_EXPORT gnPresentationQueueState gnPresentationQueueGetStateFn(gnPresentationQueue& presentationQueue) {
    return GN_VALID;
}

GN_EXPORT gnReturnCode gnCreatePresentationQueueFn(gnPresentationQueue* presentationQueue, const gnOutputDevice& device, gnPresentationDetails& details) {
    presentationQueue->presentationQueue = new gnPlatformPresentationQueue();
    presentationQueue->presentationQueue->outputDevice = const_cast<gnOutputDevice*>(&device);

    presentationQueue->presentationQueue->layer = CA::MetalLayer::layer();
    presentationQueue->presentationQueue->layer->setPixelFormat(MTL::PixelFormat::PixelFormatBGRA8Unorm_sRGB);
    presentationQueue->presentationQueue->layer->setFramebufferOnly(true);
    presentationQueue->presentationQueue->layer->setDrawableSize({ (double)details.ImageSize.x, (double)details.ImageSize.y });
    presentationQueue->presentationQueue->layer->setDevice(device.outputDevice->device);

    mtlInitializeMetalLayer(presentationQueue->presentationQueue->layer, true);

    mtlObjectCSetContentViewsLayer(
        device.outputDevice->instance->instance->window,
        presentationQueue->presentationQueue->layer);

    for (int i = 0; i < details.ImageCount; i++) {
        MTL::TextureDescriptor* desc = MTL::TextureDescriptor::texture2DDescriptor(
            MTL::PixelFormat::PixelFormatBGRA8Unorm,
            details.ImageSize.x,
            details.ImageSize.y,
            false
        );
        desc->setUsage(MTL::TextureUsageRenderTarget | MTL::TextureUsageShaderRead);

        gnTexture newImage = gnTexture();
        newImage.textureColorFormat = GN_BGRA8;
        newImage.minFilter = GN_FILTER_NEAREST;
        newImage.magFilter = GN_FILTER_NEAREST;
        newImage.textureType = GN_TEXTURE_2D;
        newImage.texture = new gnPlatformTexture();
        newImage.texture->texture = device.outputDevice->device->newTexture(desc);
        gnListAdd(presentationQueue->images, newImage);
    }

    MTL::TextureDescriptor* afsdfsdf = MTL::TextureDescriptor::texture2DDescriptor(
        MTL::PixelFormat::PixelFormatBGRA8Unorm,
        2,
        2,
        false
    );
    afsdfsdf->setUsage(MTL::TextureUsageRenderTarget | MTL::TextureUsageShaderRead);

    gnTexture newImage = gnTexture();
    newImage.textureColorFormat = GN_BGRA8;
    newImage.minFilter = GN_FILTER_NEAREST;
    newImage.magFilter = GN_FILTER_NEAREST;
    newImage.textureType = GN_TEXTURE_2D;
    newImage.texture = new gnPlatformTexture();
    newImage.texture->texture = device.outputDevice->device->newTexture(afsdfsdf);
    MTL::Region region = MTL::Region(0, 0, 2, 2);
    uint32_t data[4] = {
        0xffffffff, 0xff000000, 0xffff0000, 0xffffff00
    };
    newImage.texture->texture->replaceRegion(region, 0, data, 8);

    gnListAdd(presentationQueue->images, newImage);

    // so funny story I have no god damn clue how to implement a presentation queue on metal
    return GN_SUCCESS;
}
GN_EXPORT void gnDestroyPresentationQueueFn(gnPresentationQueue& queue) {
    for (int i = 0; i < gnListLength(queue.images); i++) {
        gnDestroyTexture(queue.images[i]);
    }
    queue.images = gnCreateList<gnTexture>();
}

File: rendering_api/metal/src/core/presentation_queue/metal_presentation_queue.m
#include "metal_presentation_queue.h"
#include "core/surface/metal_surface.h"
#include "core/devices/metal_output_devices.h"
#include "core/debugger/gryphn_debugger.h"
#include "core/texture/metal_texture.h"
#include "core/sync/semaphore/metal_semaphore.h"

gnReturnCode gnCreatePresentationQueueFn(gnPresentationQueue* presentationQueue, const gnOutputDevice* device, struct gnPresentationQueueInfo_t presentationInfo) {
    if (presentationInfo.minImageCount > 3) {
        gnDebuggerSetErrorMessage(device->instance->debugger, (gnMessageData){
            .message = gnCreateString("On Metal you cannot have more than 3 images in a presentation queue")
        });
        return GN_UNSUPPORTED_IMAGE_COUNT;
    }

    if (presentationInfo.minImageCount < 2) {
        gnDebuggerSetErrorMessage(device->instance->debugger, (gnMessageData){
            .message = gnCreateString("On Metal you cannot have less than 2 images in a presentation queue")
        });
        return GN_UNSUPPORTED_IMAGE_COUNT;
    }

    presentationQueue->presentationQueue = malloc(sizeof(struct gnPlatformPresentationQueue_t));

    MTLPixelFormat convertedFormat = mtlGryphnFormatToVulkanFormat(presentationInfo.format.format);
    CGColorSpaceRef convertedColorSpace = mtlGryphnColorSpaceToVulkanColorSpace(presentationInfo.format.colorSpace);

    presentationQueue->presentationQueue->textureCount = presentationInfo.minImageCount;
    presentationQueue->presentationQueue->textures = malloc(sizeof(id<MTLTexture>) * presentationInfo.minImageCount);

    MTLTextureDescriptor* textureDescriptor = [[MTLTextureDescriptor alloc] init];
    textureDescriptor.pixelFormat = convertedFormat;
    textureDescriptor.width = presentationInfo.imageSize.x;
    textureDescriptor.height = presentationInfo.imageSize.y;
    textureDescriptor.usage = MTLTextureUsageRenderTarget;
    textureDescriptor.textureType  = MTLTextureType2D;

    presentationQueue->imageCount = presentationInfo.minImageCount;
    presentationQueue->images = malloc(sizeof(gnTexture) * presentationInfo.minImageCount);
    for (int i = 0; i < presentationInfo.minImageCount; i++) {
        presentationQueue->presentationQueue->textures[i] = [device->outputDevice->device newTextureWithDescriptor:textureDescriptor];
        presentationQueue->images[i].texture = malloc(sizeof(gnPlatformTexture));
        presentationQueue->images[i].texture->texture = presentationQueue->presentationQueue->textures[i];
    }

    return GN_SUCCESS;
}

gnReturnCode gnPresentationQueueGetImageFn(gnPresentationQueue* presentationQueue, uint64_t timeout, struct gnSemaphore_t* semaphore, uint32_t* imageIndex) {
    semaphore->semaphore->eventTriggered = gnFalse;
    *imageIndex = presentationQueue->presentationQueue->currentImage;
    presentationQueue->presentationQueue->currentImage++;
    presentationQueue->presentationQueue->currentImage %= presentationQueue->imageCount;
    semaphore->semaphore->eventTriggered = gnTrue;
    return GN_SUCCESS;
}

void gnDestroyPresentationQueueFn(gnPresentationQueue *presentationQueue) {
    for (int i = 0; i < presentationQueue->imageCount; i++) {
        [presentationQueue->presentationQueue->textures[i] release];
    }
    free(presentationQueue->presentationQueue);
}
